tests/filecheck/transforms/convert_onnx_to_linalg.mlir

```diff
@@ -30,13 +30,42 @@
 
-// CHECK-NEXT:   %t5, %t6, %t7 = "test.op"() : () -> (tensor<1x320xf32>, tensor<50x320xf32>, tensor<50xf32>)
+// CHECK-NEXT:  %t5, %t6, %t7 = "test.op"() : () -> (tensor<1x320xf32>, tensor<50x320xf32>, tensor<50xf32>)
 // CHECK-NEXT:  %3 = tensor.empty() : tensor<320x50xf32>
 // CHECK-NEXT:  %4 = linalg.transpose ins(%t6:tensor<50x320xf32>) outs(%3:tensor<320x50xf32>) permutation = [1, 0]
 // CHECK-NEXT:  %res_gemm = tensor.empty() : tensor<1x50xf32>
 // CHECK-NEXT:  %res_gemm_1 = linalg.mul ins(%t5, %4 : tensor<1x320xf32>, tensor<320x50xf32>) outs(%res_gemm : tensor<1x50xf32>) -> tensor<1x50xf32>
 // CHECK-NEXT:  %res_gemm_2 = linalg.add ins(%res_gemm_1, %t7 : tensor<1x50xf32>, tensor<50xf32>) outs(%res_gemm : tensor<1x50xf32>) -> tensor<1x50xf32>
 
+
+%t8, %t9, %t10 = "test.op"(): () -> (tensor<5x3xf32>, tensor<3x2xf32>, tensor<5x2xf32>)
+%res_gemm_1 = "onnx.Gemm"(%t8, %t9, %t10) {onnx_node_name = "/Gemm"}: (tensor<5x3xf32>, tensor<3x2xf32>, tensor<5x2xf32>) -> tensor<5x2xf32>
+
+
+// CHECK-NEXT:  %t8, %t9, %t10 = "test.op"() : () -> (tensor<5x3xf32>, tensor<3x2xf32>, tensor<5x2xf32>)
+// CHECK-NEXT:  %res_gemm_1 = tensor.empty() : tensor<5x2xf32>
+// CHECK-NEXT:  %res_gemm_1_1 = linalg.mul ins(%t8, %t9 : tensor<5x3xf32>, tensor<3x2xf32>) outs(%res_gemm_1 : tensor<5x2xf32>) -> tensor<5x2xf32>
+// CHECK-NEXT:  %res_gemm_1_2 = linalg.add ins(%res_gemm_1_1, %t10 : tensor<5x2xf32>, tensor<5x2xf32>) outs(%res_gemm_1 : tensor<5x2xf32>) -> tensor<5x2xf32>
+
+
+%t11, %t12, %t13 = "test.op"(): () -> (tensor<10x5xf32>, tensor<10x3xf32>, tensor<5x3xf32>)
+%res_gemm_2 = "onnx.Gemm"(%t11, %t12, %t13) {onnx_node_name = "/Gemm", "alpha" = 0.500000e+00 : f32, "beta" = 0.500000e+00 : f32, "transA" = 1 : si64}: (tensor<10x5xf32>, tensor<10x3xf32>, tensor<5x3xf32>) -> tensor<5x3xf32>
+
+
+// CHECK-NEXT:  %t11, %t12, %t13 = "test.op"() : () ->  (tensor<10x5xf32>, tensor<10x3xf32>, tensor<5x3xf32>)
+// CHECK-NEXT:  %5 = tensor.empty() : tensor<5x10xf32>
+// CHECK-NEXT:  %6 = linalg.transpose ins(%t11:tensor<10x5xf32>) outs(%5:tensor<5x10xf32>) permutation = [1, 0]
+// CHECK-NEXT:  %7 = tensor.empty() : tensor<5x10xf32>
+// CHECK-NEXT:  %8 = arith.constant 5.000000e-01 : f32
+// CHECK-NEXT:  %9 = linalg.mul ins(%8, %6 : f32, tensor<5x10xf32>) outs(%7 : tensor<5x10xf32>) -> tensor<5x10xf32>
+// CHECK-NEXT:  %10 = tensor.empty() : tensor<5x3xf32>
+// CHECK-NEXT:  %11 = arith.constant 5.000000e-01 : f32
+// CHECK-NEXT:  %12 = linalg.mul ins(%11, %t13 : f32, tensor<5x3xf32>) outs(%10 : tensor<5x3xf32>) -> tensor<5x3xf32>
+// CHECK-NEXT:  %res_gemm_2 = tensor.empty() : tensor<5x3xf32>
+// CHECK-NEXT:  %res_gemm_2_1 = linalg.mul ins(%9, %t12 : tensor<5x10xf32>, tensor<10x3xf32>) outs(%res_gemm_2 : tensor<5x3xf32>) -> tensor<5x3xf32>
+// CHECK-NEXT:  %res_gemm_2_2 = linalg.add ins(%res_gemm_2_1, %12 : tensor<5x3xf32>, tensor<5x3xf32>) outs(%res_gemm_2 : tensor<5x3xf32>) -> tensor<5x3xf32>
```

Comment:
This looks weird to me. My understanding is that the result should be `A @ B * alpha + beta` (modulo transpose). Is that the case? If so, the values for each subexpression (`A @ B`, `A @ B * alpha`, and the result) all have the same shape, and can be stored in the same result buffer, so with only one tensor.empty. Something like this:
```
res0 = tensor.empty()
res1 = linalg.mul(A,B, out=res0)
alpha = arith.constant alpha
res2 = linalg.mul(res1, alpha, out=res1)
beta = arith.constant beta
res3 = linalg.add(res2, beta, out=res2)
```
I'm not 100% sure that this is legal with the tensor semantics, but this is the shape that we'd want with memrefs.

Reply:
@PapyChacal, Mr Bufferization, do you have advice for this? Is it legal to pass a tensor as both an in and out parameter? is the tensor.empty the way to go here?

Reply:
I checked and it seems fine to do this

Reply:
@superlopuh  does this mean we are good to merge this ?

Reply:
👍 thanks for the reminder